back/Dockerfile

```diff
@@ -1,31 +1,57 @@
-FROM ruby:2.7.6-slim
+# The purpose of this build stage is to extract only the dependency files (gemfiles and
+# gemspec files) from the project. This enables us to optimize Docker layer caching by
+# installing all the dependencies before copying the remaining project files to the
+# Docker image in the later stages.
+FROM alpine:3.17.2 AS dependency_files
+SHELL ["/bin/ash", "-o", "pipefail", "-c"]
 
-ARG CITIZENLAB_EE=false
-ENV CITIZENLAB_EE=$CITIZENLAB_EE
+WORKDIR /app
 
-RUN apt-get update && apt-get install -qq -y --no-install-recommends \
-      build-essential libpq-dev file imagemagick curl git optipng jpegoptim pngquant libgeos-dev libgmp3-dev netcat shared-mime-info \
-      less
+COPY back/Gemfile back/Gemfile.* ./
+COPY back/engines ./engines
 
-RUN curl -sL https://deb.nodesource.com/setup_15.x  | bash -
-RUN apt-get -y install nodejs
+# `version.rb` files are also kept as they are required by gemspec files.
+RUN find engines -type f -not -name "*.gemspec" -not -name "version.rb" -print0 | xargs -0 rm -rf \
```

Comment:
I have no idea, but do we necessarily need different stages for it?
Could we just do sth like
```Dockerfile
COPY back/Gemfile back/Gemfile.* ./
COPY back/engines ./engines

# `version.rb` files are also kept as they are required by gemspec files.
RUN find engines -type f -not -name "*.gemspec" -not -name "version.rb" -print0 | xargs -0 rm -rf \
&& find engines -type d -empty -delete

RUN bundle install
COPY back/. .
```

Reply:
Ah yes, it makes sense. Thanks for explaining!

Reply:
Ah, I think I get it now :) We're always executing that first stage, but we're not bundling in it, just using it to get the right files out, so it's really fast. Right?